packages/world/src/systems/HitPlayerSystem.sol

```diff
@@ -0,0 +1,94 @@
+
+import { EntityId } from "../types/EntityId.sol";
+import { ObjectTypes } from "../types/ObjectType.sol";
+import { ProgramId } from "../types/ProgramId.sol";
+import { Vec3 } from "../types/Vec3.sol";
+import { RateLimitUtils } from "../utils/RateLimitUtils.sol";
+
+contract HitPlayerSystem is System {
+  function hitPlayer(EntityId caller, EntityId target, uint16 toolSlot, bytes calldata extraData) public {
+    _hitPlayer(caller, target, toolSlot, extraData);
+  }
+
+  function hitPlayer(EntityId caller, EntityId target, bytes calldata extraData) public {
+    _hitPlayer(caller, target, type(uint16).max, extraData);
+  }
+
+  function _hitPlayer(EntityId caller, EntityId target, uint16 toolSlot, bytes calldata extraData) internal {
+    // Update and check caller's energy
+    caller.activate();
+
+    (, Vec3 targetCoord) = caller.requireInRange(target, MAX_HIT_RADIUS);
+
+    require(target != caller, "Cannot hit yourself");
+    require(target._exists(), "No entity at target location");
+    require(target._getObjectType() == ObjectTypes.Player, "Target is not a player");
+
+    // Check rate limit for combat actions
+    RateLimitUtils.hit(caller);
+
+    // Update target player's energy
+    uint128 targetEnergy = updatePlayerEnergy(target).energy;
+
+    // If target is already dead, return early
+    if (targetEnergy == 0) {
+      return;
+    }
+
+    // Get tool data for damage calculation
+    ToolData memory toolData = ToolUtils.getToolData(caller, toolSlot);
+
+    // Use tool and get total damage (handles energy costs internally)
+    uint128 damage = toolData.use(targetEnergy, HIT_ACTION_MODIFIER, toolData.toolType.isWhacker());
+
+    // If caller died (damage == 0), return early
+    if (damage == 0) {
+      return;
+    }
+
+    // Apply damage to target player
+    decreasePlayerEnergy(target, targetCoord, damage);
+
+    // Add target's total damage to target's local pool
+    addEnergyToLocalPool(targetCoord, damage);
+
+    _requireHitsAllowed(caller, targetCoord, toolData.tool, damage, extraData);
+
+    // Notify both players about the hit
+    notify(caller, HitPlayerNotification({ targetPlayer: target, targetCoord: targetCoord, damage: damage }));
+
+    // TODO: Notify target about being hit?
+    // notify(target, HitPlayerNotification({ targetPlayer: caller, targetCoord: callerCoord, damage: totalDamage }));
+  }
+
+  function _requireHitsAllowed(
+    EntityId caller,
+    Vec3 targetCoord,
+    EntityId tool,
+    uint128 totalDamage,
+    bytes calldata extraData
+  ) internal {
+    // Check if target is within a force field and call hooks
+    (ProgramId program, EntityId hookTarget, EnergyData memory energyData) = ForceFieldUtils.getHookTarget(targetCoord);
+
+    if (!program.exists()) {
+      return;
+    }
+
+    program.hook({ caller: caller, target: hookTarget, revertOnFailure: energyData.energy > 0, extraData: extraData })
+      .onHit(tool, totalDamage);
```

Comment:
hmm we're using this for hitting the machine right? How would the program know this is a hit on the player?

Reply:
yes I think we will need to add a new field or have a different hook like onHitPlayer

Reply:
nice so its now a field on onHit! this will require a migration of existing force field programs right?